game/src/plugins/skills/fishing/src/fishing.plugin.kts

```diff
@@ -0,0 +1,133 @@
+import org.apollo.cache.def.ItemDefinition
+import org.apollo.game.action.DistancedAction
+import org.apollo.game.message.impl.NpcActionMessage
+import org.apollo.game.model.Position
+import org.apollo.game.model.entity.Player
+import org.apollo.game.model.entity.Skill
+import org.apollo.game.plugin.skills.fishing.*
+import kotlin.properties.Delegates
+import java.util.Random
+
+class FishingAction: DistancedAction<Player> {
+
+    var spot: FishingSpot by Delegates.notNull()
+    var tool: FishingTool by Delegates.notNull()
+    var started: Boolean = false
+    var options: IntArray by Delegates.notNull()
+    var minLevel: Int = 110
+
+    constructor(mob: Player, position: Position, spot: FishingSpot, option: Int) : super(4, true, mob, position, 1) {
+        this.tool = FISHING_TOOLS[spot.tools[option - 1]]!!
+        this.spot = spot
+        this.options = if (option == 1) spot.firstFish else spot.secondFish
+        //Get min level in really compact way
+        options
+                .asSequence()
+                .filter { minLevel > FISH[it]!!.level }
+                .forEach { minLevel = FISH[it]!!.level }
+        //
+    }
+
+    fun successfulCatch(level: Int, req: Int): Boolean {
+        return intArrayOf(level - req + 5, 30).min()!! > Random().nextInt(40)
```

Comment:
We shouldn't be creating new Random instances every time.  Also, is there a source on this calculation?

Reply:
OK, I can add TODO comments to the code for those, for testing I will keep the current impl in place. Also, I still have not found the original source of this particular calculation other than the ruby that it came form. Any ideas?

Reply:
I'm pretty sure I made it up on the spot when writing the plugin

Reply:
OK, thanks. That is good to know.